core/src/main/scala/kafka/log/Log.scala

```diff
@@ -744,7 +744,12 @@ class Log(@volatile var dir: File,
           s"index_size = ${segment.index.entries}/${segment.index.maxEntries}, " +
           s"time_index_size = ${segment.timeIndex.entries}/${segment.timeIndex.maxEntries}, " +
           s"inactive_time_ms = ${segment.timeWaitedForRoll(now, maxTimestampInMessages)}/${config.segmentMs - segment.rollJitterMs}).")
-      roll()
+      /*
+        maxOffsetInMessages - Integer.MAX_VALUE is a heuristic value for the first offset in the set of messages.
+        Since the offset in messages will not differ by more than Integer.MAX_VALUE, this is guaranteed <= the real
+        min offset
```

Comment:
Bucket's suggestion seems to be the best option for now. Since it's a bit hachy, could we document a bit more why we need this, especially with respect to compacted topic?